### PR TITLE
starts 'After Swap' procedure

### PR DESCRIPTION
**Pull request by rbcasperson**

### CODE REVIEW

**Comment by rbcasperson** [comment]
I got this to "work" based on what I thought it needs to do. I'm looking more into how the rest of `Board` works to see if I can integrate it more nicely.

**Comment by rbcasperson** [comment]
Okay I understand how `this.orbs` is the board. It took me a little while to figure it out.

**Comment by rbcasperson** [comment]
See commit message ^

File: babel/board.js [hunk 1]
```diff
@@ -46,12 +47,48 @@ export default class Board {
         return _.uniq(_.flatten(this.orbs)).sort();
     }
 
-    /**
-     * @returns {Object} - A match event. Includes the
-     *
-     */
+    findMatches() {
+        let matches = [],
+            b = _.cloneDeep(this.orbs);
```
Comment (Droogans) [inline babel/board.js]: Do you really need to clone the orbs?

Comment (rbcasperson) [inline babel/board.js]: I was going to ask you about cloning. In this case, I think it is just so there isn't `this.orbs` everywhere. I guess it was aesthetic here.

File: babel/board.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
 'use strict';
 import * as _ from 'lodash';
+let SortedSet = require('collections/sorted-set');
```
Comment (Droogans) [inline babel/board.js:3]: Node v5 should come with `Set` objects already. Is it just as easy as creating a set, and then sorting it?

Comment (rbcasperson) [inline babel/board.js:3]: I saw that but wasn't sure. I was looking through the collections module and in order to use `union()` it had to be a SortedSet and not a Set.

File: babel/board.js [hunk 3]
```diff
@@ -46,12 +47,48 @@ export default class Board {
...
+        });
+        return matches;
+    }
+    
+    combineMatches() {
+        let combinedMatches = [],
+            unused = _.clone(this.findMatches()), 
+            couldMatch, before, currentMatch;
+        
+        while (unused[0] != null) {
```
Comment (Droogans) [inline babel/board.js]: I think this is simpler read as

``` js
while (unused.shift()) {
    // ...
}
```

that'll break from the while loop once `shift()` returns null.

File: babel/board.js [hunk 4]
```diff
@@ -46,12 +47,48 @@ export default class Board {
...
+            currentMatch = new SortedSet(unused[0]);
+            unused.shift();
+            couldMatch = _.clone(unused);
+                
+            _.each(couldMatch, m => { //only union if there is an overlap!
+                if (currentMatch.intersection(m).toArray()[0] != null) {
```
Comment (Droogans) [inline babel/board.js]: Ah, I see now why you want to use the collections library.

Ok.

File: test/board.spec.js [hunk 5]
```diff
@@ -19,6 +19,46 @@ test('creates seven different types of orbs', t => {
...
+        [ 4, 2, 0, 5, 3, 0, 1, 5 ],
+        [ 1, 4, 3, 2, 5, 4, 6, 6 ],
+        [ 3, 5, 2, 2, 2, 3, 3, 4 ],
+        [ 1, 1, 3, 2, 5, 2, 6, 2 ],
+        [ 3, 2, 5, 4, 5, 0, 4, 0 ],
+        [ 1, 3, 1, 6, 1, 6, 3, 3 ],
+        [ 1, 0, 3, 0, 1, 1, 3, 3 ] 
+    ];
+    
+    t.same(board.combineMatches(board.findMatches()), testCombined);
```
Comment (Droogans) [inline test/board.spec.js]: Why are you calling `combineMatches` with an argument? It doesn't appear to accept one.

**Comment by Droogans** [comment]
I think a few of these functions should be standalone-functions that are exported alongside the `Board` class. Right now everything is attached to the instance that `new Board()` creates, and it's not very intuitive.

File: test/board.spec.js (hunk 5 above)
Comment (rbcasperson) [inline test/board.spec.js]: I adjusted it in `board.js` but not in the test

**Comment by Droogans** [comment]
I'm going to pull this down and rip `combineMatches` and `findMatches` into standalone functions that accept any set of orbs.

File: babel/board.js [hunk 6]
```diff
@@ -46,12 +47,48 @@ export default class Board {
...
+                };
+            });
+        });
+        return matches;
+    }
+    
+    combineMatches() {
+        let combinedMatches = [],
+            unused = _.clone(this.findMatches()), 
+            couldMatch, before, currentMatch;
```
Comment (Droogans) [inline babel/board.js]: I personally feel strongly against using multiline, comma separated `let` definitions with one `let` statement.

It's better to use one let, one assignment. That way, you can move them around, cut/paste them, delete them, etc., without worrying about `let` being in the proper place.

``` js
let combinedMatches = [];
let unused = _.clone(this.findMatches());
// ...
```

**Comment by Droogans** [comment]
I just finished https://github.com/Droogans/match-three-js/commit/05adc89515e3ee48baf3a8c45f6dd66222e0881a I would want to see this added before merging.

**Comment by Droogans** [comment]
Alright, I'm still not 100% on this PR, but it's good enough to at least iterate on.

I'll start by cleaning up my old code first, then I'll add a few more tests to the functions you've introduced here. I think after a couple more PRs, the basics will be finally complete.

**Pull request merged**